src/assistant.ts

```diff
@@ -24,6 +25,8 @@ export type AppHandler = OmniHandler & BaseApp
 export interface AppOptions {
   /** @public */
   debug?: boolean
+  /** @public */
+  logger?: Logger
```

Comment:
Is logging something that could be implemented outside the library with a middleware plugin?

Reply:
Feels more like an `app.use` plugin than an `app.middleware` plugin to me, if anything. The call to `use` could then cascade a call to `middleware` to ensure that the logger is injected into each conversation as needed.

The difficult part is deciding what logging API(s) to support and how to get the rest of the SDK to use the same logger as the user.

Reply:
In this PR, there is a new Logger module which basically reimplements methods from the common module. Is that set of methods sufficient? If not, how much does it matter seeing as you're implementing the logging in your own domain? If you really want a new log method, you could add it to the plugin and use that method directly in your codebase.

Reply:
@Fleker @igilham Thank you for the comments!

First of all, current log outputs (`Request`, `Response`, `Conversation` and `Headers`) are very valuable for most developers to know what was do inside the SDK and in the code set the developers wrote. Also, the developers can get their log outputs without any coding. Furthermore, the developers currently can't access to the objects (of course, they are the request, response, conv and headers, especially, the response object) from the code they wrote directly.

Generally, many developers are watching the log outputs during developing their actions. But, there are valuable information in the log outputs for an operation and a maintenance of the actions. If the log outputs is not as plain text, but they are structured data, developers can re-use the information for the operation and the maintenance. For example, developers can get the information from pub/sub and function. Also, developers can export them to BigQuery directly. Currently, developers can't do them easily.

And, in Azure Functions, the `context.log` is generally used instead of `console.log`. However, Azure developers can't replace `console.log` with `context.log` because of hardcoding in the `common` module.

> Feels more like an `app.use` plugin than an `app.middleware` plugin to me, if anything.

Yes, actually the `app.use` is a suitable solution, I think, too. But,  it seems that the purpose of the current `app.use` plugin mechanism is to add/override methods to the `app` object. And, there is no API to replace the logging logic. I added the interface to `AppOptions`, but, for example, I have other idea which I add the API to `ServiceBaseApp` interface (ex. `setLogger`). So, developers may provide own plugin to replace to own logger object with the API.